scripts/release

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env ruby
+
+require "date"
+require "optparse"
+
+def run(command, abort_on_fail: true)
+  puts "> #{command}" if $VERBOSE
+
+  system(command, exception: abort_on_fail)
+rescue => e
+  raise "#{command.inspect} failed: #{e}"
+end
+
+channel = ENV.fetch("FOREM_RELEASE_CHANNEL", "stable")
+date = ENV.fetch("FOREM_RELEASE_DATE", Date.today.to_s).gsub(/-/, "")
+hotfix = ENV.fetch("FOREM_RELEASE_HOTFIX", 0)
+remote = ENV.fetch("FOREM_RELEASE_REMOTE", "origin")
+source = ENV.fetch("FOREM_STAGING_FROM", "#{channel}-next")
+
+OptionParser.new ARGV.dup do |options|
+  options.banner = <<~USAGE
+    Usage: #$0 [options]
+
+    Releases are named CHANNEL.DATE.HOTFIX_COUNT. Example: #{channel}.#{date}.#{hotfix}
+
+  USAGE
+
+  options.on "-c", "--channel CHANNEL", "Specify the release channel (defaults to #{channel.inspect})" do |release_channel|
+    channel = release_channel
+  end
+
+  options.on "-d", "--date DATE", "Specify the release date (defaults to today)" do |release_date|
+    date = release_date.gsub(/-/, "")
+  end
+
+  options.on "-f", "--from BRANCH/COMMIT/TAG", "Specify the source branch for this release (defaults to #{source.inspect})" do |from_branch|
+    source = from_branch
+  end
+
+  options.on "-h", "--hotfix HOTFIX", "Specify the hotfix count for this release branch (defaults to #{hotfix.inspect})" do |hotfix_opt|
+    hotfix = hotfix_opt
+  end
+
+  options.on "-r", "--remote REMOTE", "Git remote to push to (defaults to #{remote.inspect})" do |git_remote|
+    remote = git_remote
+  end
+
+  options.on "-v", "--verbose", "Enable verbose mode" do
+    $VERBOSE = true
+  end
+end.parse!
+
+original_branch = `git branch --show-current`.strip
+
+begin
+  tag = "#{channel}.#{date}.#{hotfix}"
+  puts "Building #{tag}..."
+
+  [source, channel].each do |branch|
+    run "git checkout --quiet #{branch}"
```

Comment:
Since we are trying to fast-forward we should be able to use `git fetch <remote> <sourceBranch>:<destinationBranch>` . This would allow us to save two checkouts and the `ensure` block at the end.

https://stackoverflow.com/questions/3216360/merge-update-and-pull-git-branches-without-using-checkouts

Reply:
> This would allow us to save two checkouts and the `ensure` block at the end.

Are there other benefits to avoiding checkouts beyond not needing to reset the checkout state? 

For a little more background into why I chose this approach: with the `branch1:branch2` format, I'm never sure if the branches are meant to be `local:remote` vs `source:destination`, or if either of those are even the correct order. I could look it up, but experience tells me that I will not retain this information and would need to look it up again every time I see it[1]. :joy: 

`git checkout` and `git pull` are pretty easily understood, though, making this form really easy to grok without clarifying comments.

---

[1] Same for the Unix `ln` command, Minitest's `assert_equal`, the x86 `mov` instruction, etc. I have to look these up every single time. 😆 I'm just so bad with positional arguments in general.

Reply:
No other benefits I can think of. I just thought if we can turn something that is a two-step process that requires a "cleanup" step into a one-step process that doesn't it's at least worth considering. I'm totally fine with leaving it as-is though.